service/src/main/java/bio/terra/drshub/services/SignedUrlService.java

```diff
@@ -0,0 +1,54 @@
+package bio.terra.drshub.services;
+
+import bio.terra.bond.model.SaKeyObject;
+import bio.terra.common.iam.BearerToken;
+import bio.terra.drshub.config.DrsHubConfig;
+import bio.terra.drshub.logging.AuditLogEvent;
+import bio.terra.drshub.logging.AuditLogEventType;
+import bio.terra.drshub.logging.AuditLogger;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import java.net.URL;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+public record SignedUrlService(
+    DrsHubConfig drsHubConfig,
+    AuthService authService,
+    DrsProviderService drsProviderService,
+    GoogleStorageService googleStorageService,
+    AuditLogger auditLogger) {
+
+  public URL getSignedUrl(
+      String bucket,
+      String objectName,
+      String dataObjectUri,
+      String googleProject,
+      BearerToken bearerToken,
+      String ip) {
+
+    var components = drsProviderService.getUriComponents(dataObjectUri);
+    var drsProvider = drsProviderService.determineDrsProvider(components);
+    SaKeyObject saKey = authService.fetchUserServiceAccount(drsProvider, bearerToken);
+    Storage storage = googleStorageService.getAuthedStorage(saKey, googleProject);
+    BlobInfo blobInfo = BlobInfo.newBuilder(BlobId.of(bucket, objectName)).build();
+    var duration = drsHubConfig.getSignedUrlDuration();
+
+    var logEvent =
+        new AuditLogEvent.Builder()
+            .dRSUrl(dataObjectUri)
```

Comment:
What about the `project/bucket/object`?   We're logging the DRS URI but not the bucket/object that we are _actually_ signing.

Reply:
The `AuditLogEvent` doesn't have a field for that, and I'm hesitant to add a field for fear of messing up some logging config downstream. The `drsUrl` should be enough for audit logging, since we can always resolve the object from that. Plus, the actual accessing of the gs url will be logged when the signed url is used.

Reply:
Well, not necessarily, we could send a DRS URI that doesn't point at the same project/bucket/object I think.  I don't recall seeing anything that enforced those matching data location of the DRS URI.